port/linux/ipadapter.c

```diff
@@ -899,15 +892,119 @@ process_event(ip_context_t *dev, fd_set *rdfds, fd_set *wfds)
   return 0;
 }
 
+#ifdef OC_DYNAMIC_ALLOCATION
+static bool
+fd_sets_are_equal(const fd_set *fd1, const fd_set *fd2)
+{
+#ifdef __FDS_BITS
+  return (memcmp(__FDS_BITS(fd1), __FDS_BITS(fd2), sizeof(__FDS_BITS(fd1))) ==
+          0);
+#else  //!__FDS_BITS
+  return (memcmp(fd1->fds_bits, fd2->fds_bits, sizeof(fd1->fds_bits)) == 0);
```

Comment:
This might not be portable, as far as I understand it, POSIX doesn't define how the fd_set struct looks, so the member might be called `fds_bits`. What is guaranteed to exist are the `FD_*` macros. So I guess the most portable thing we can do a for-cycle in the general case:

```C
#ifdef __FDS_BITS
  return (memcmp(__FDS_BITS(fd1), __FDS_BITS(fd2), sizeof(__FDS_BITS(fd1))) ==
          0);
#else  // !__FDS_BITS
  for (int i=0; i<FD_SETSIZE; ++i) {
    if (FD_ISSET(i, fd1) != FD_ISSET(i, fd2)) {
      return false;
    }
  }
  return true;
#endif // __FDS_BITS
```